File: tree/binaryTree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *left;
    struct Node *right;
} Node;

Node *createNode(int value) {
    Node *newNode = (Node*)malloc(sizeof(Node));
    newNode->data = value;
    newNode->right = NULL;
    newNode->left = NULL;

    return newNode;
}

Node *insert(Node *root, int value) {
    if (root == NULL) return createNode(value);

    if (value < root->data) {
        root->left = insert(root->left, value);
    }
    if (value > root->data) {
        root->right = insert(root->right, value);
    }

    return root;
}

void preOrder(Node *root) {
    if (root != NULL) {
        printf("%d -> ", root->data);
        preOrder(root->left);  
        preOrder(root->right);
    }
}

void order(Node *root) {
    if (root != NULL) {
        order(root->left);
        printf("%d -> ", root->data);
        order(root->right);
    }
    
}

int main() {
    Node *root = NULL;
    int n;
    
    root = insert(root, 20);

    for (int i = 0; i < 5; i++) {
        scanf("%d", &n);
        root = insert(root, n); 
    }


    preOrder(root);
    printf("\n");
    order(root);
    return 0;
}

File: tree/binaryTree.py
class treeNode:
    def __init__(self,data):
        self.data = data
        self.left = None
        self.right = None

class binaryTree:
    def __init__(self):
        self.root = None

    def insert(self,data):
        if not self.root:
            self.root = treeNode(data)
        else:
            self._insert_recursive(self.root,data)

    def _insert_recursive(self,node,data):
        if data < node.data:
            if node.left:
                self._insert_recursive(node.left, data)
            else:
                node.left = treeNode(data)

        else:
            if node.right:
                self._insert_recursive(node.right, data)
            else:
                node.right = treeNode(data)

    def search(self,data):
        return self._search_recursive(self.root,data)

    def _search_recursive(self,node,data):
        if not node:
            return False

        if node.data == data:
            return True
        if node.data < data:
            return self._search_recursive(node.right,data)
        return self._search_recursive(node.left,data)


tree = binaryTree()
root = treeNode(20)

values = [10, 30, 5, 15, 25, 35]
tree.insert(root.data)
for value in values:
    tree.insert(value)

print(tree.search(20))  # Output: True
print(tree.search(100))  # Output: False 
